lib/fennec/stun/auth.ex

```diff
@@ -0,0 +1,66 @@
+defmodule Fennec.STUN.Auth do
+  @moduledoc """
+  This module implements authentication and authorization of the STUN protocol.
+  """
+
+  alias Jerboa.Format.Body.Attribute
+  alias Jerboa.Format.Body.Attribute.{Username, Nonce, Realm}
+  alias Jerboa.Params
+
+  def get_secret do
+    Application.get_env(:fennec, :secret)
+  end
+
+  def authorize(params, _turn_state) do
+    with u = %Username{} <- Params.get_attr(params, Username),
+         r = %Realm{}    <- Params.get_attr(params, Realm) do
+      # Right now, any authenticated user will do
+      {:ok, %Params{params | attributes: params.attributes -- [u, r]}}
+    end
+  end
+
+  def authenticate(params, turn_state) do
+    nonce = turn_state.nonce
+    signed? = params.signed?
+
+    with %Username{}                <- Params.get_attr(params, Username),
+         %Realm{}                   <- Params.get_attr(params, Realm),
+         n = %Nonce{value: ^nonce}  <- Params.get_attr(params, Nonce),
+         true                       <- params.verified? do
+      {:ok, %Params{params | attributes: params.attributes -- [n]}}
+    else
+      false -> # Not verified -> error code 401
+        {:error, error_params(401, params, turn_state)}
+      _ when signed? -> # If message is signed and there are some attributes
+                        # missing, we need to respond with error code 400
+        {:error, error_params(400, params, turn_state)}
+      %Nonce{} -> # Invalid nonce, error code 438
+        {:error, error_params(438, params, turn_state)}
+      nil when not signed? ->
+        {:error, error_params(401, params, turn_state)}
+    end
+  end
+
+  def maybe(action_fun, params, turn_state) do
+    case should_authorize?(params) do
```

Comment:
It could, but I don't like it. It's handy but only if there is no `else` while here we would need it.